W-14079120-firstusecaseexample-se (#296)
mulesoft/docs-rpa: rpa-home/modules/ROOT/pages/automation-usecase-legacysystemdataextraction.adoc

# Usecase: Extract Data from a Legacy System

> Extracting data from legacy systems is a common use case for RPA. 

Reviewer: Suggest giving some examples of when you want to extract data from a legacy system - and what you might do with it.

Author: done

This example illustrates the following tasks:

* Copy the contents from a PDF file to a text file.
* Compare strings.
* Copy a substring from a text file.
* Find text on a screen using optical character recognition (OCR).

> * Interact with a screen using mouse actions and key strokes.

Reviewer: ```suggestion
* Interact with a screen using mouse actions and keystrokes.
```

* Use activity parameters to transfer data and to determine which path to follow in the model.

## Example Scenario

> Imagine you work as an accountant at Mythical University. Mythical University has ordered a travel package for one of their field excursions. You have received an email and saved the attached order confirmation to your computer.

Reviewer: Does SF use this construction (Imagine...)?

You could do something like:
Mythical University orders a travel package for a field excursion.
As an accountant at...., you receive an email confirmation and save the attached order to your computer...

> This is what an order confirmation of your travel agency Wonderlust Travel typically looks like:

Reviewer: ```suggestion
An order confirmation from your travel agency Wonderlust Travel typically looks like this:
```

Reviewer: (Avoid "this" as the subject.)

[[confirmationletter-image]]

> image::usecase-1-wonderlusttravels-orderconfirmation-pdf.png["Letter containing a logo, headers, text and tables. The order ID is contained in a table in a header in the middle of the page. The label says 'Order ID'.", 75%]

Reviewer: Is there a way to streamline the alt text into a single sentence fragment?
"Order confirmation letter from Wanderlust Travel, listing the Order ID and Customer ID"

Author: I would like to emphasize that the needed information is hard to find at first glance.

Reviewer: I see - you could say something like:
"Order confirmation letter from Wanderlust Travel, listing the Order ID and Customer ID in a table in the middle of the page"

I think one suggestion for alt text is to focus on what's important/relevant in the image.

But maybe this is moot if non-sighted people would never use this?

Reviewer: It's ultimately up to you. I'm just giving suggestions.

> You need to search the order ID in your order tracking system to get the corresponding tracking number, which you need to reference in your payment.

Reviewer: Can you put the user value first? To reference the tracking number in your payment, you need the... so you search...

Author: Like this?

Reviewer: I can't tell what changed. 

I was thinking that the user needs the tracking number and to get it they search. So, the user task is to get the tracking number.

Reviewer: The changes in the slack conversation look great to me!

> Unfortunately your order tracking system is very old and accountants can only access it by typing in and reading from a terminal screen:

Reviewer: ```suggestion
Unfortunately, your order tracking system is old and you can access it only by typing in and reading from a terminal screen:
```

[[trackingsystem-allentries-image]]
image::as400ordertrackingscreen.png["Legacy order tracking system showing green glowing text in a terminal. Above a list of orders, there is an entry field for the order id labeled 'Order-ID' and a button labeled 'Search' next to it.", 75%]

> You click the search field, enter the order ID, and click 'Search'.

Reviewer: Should Search be bold?

Is the field named "Order-ID"?

Author: I didn't bold the ui elememts of the example application to differentiate between the example and our product. In this example application the field is indeed named 'Order-ID'

Reviewer: I see now - you're not referring to the field in this case - so ignore that comment.

On line 30, Search seems to be tagged italic.

Author: This is strange. In this preview it looks like italics but in the code it's not...

> You click the search field, enter the order ID, and click 'Search'.

Reviewer: Should this be a numbered step?

Author: No, because it's not a step to solve the problem, but a step a user without RPA needs to do (part of the example process description)

Reviewer: OK, I guess I'm not super clear on why the distinction. I mean, the user has to go through a process either way. Are you saying that you use the numbered list to distinguish the RPA way from the manual way?

Author: Yes.

[[trackingsystem-noentries-image]]
image::as400ordertrackingscreen-searchresult-ordernotfound.png["Legacy order tracking system shows no search results.", 75%]

> The list of orders is empty now. This could either mean that the order is'nt there or that you mistyped the number. Let's assume that the latter is the case. You try again and the list shows the corresponding entry:

Reviewer: ```suggestion
The list of orders is now empty, which means that either the order isn't there or you mistyped the number. Let's assume that the latter is the case. You try again and the list shows the corresponding entry:
```

Reviewer: Or you could say "The search results in no orders" or something like that.

Reviewer: Is there any way to trim the images so they don't take up so much space with just the black terminal? This will make it easier to scan, too.

Author: That's the point to prove ;) Scanning those terminals is not easy....

Author: ... for a human.

Author: As a non-native speaker, I find  "The search results in no orders" difficult to understand because I waited for the verb here at first. Like in  "The search results in no orders decided to go to the beach."

Author: Or maybe  "The search results in no orders make the user feel uneasy."

[[trackingsystem-resultentry-image]]
image::as400ordertrackingscreen-searchresult-orderfound.png["Legacy order tracking system shows one search result.", 75%]

Now you can copy the tracking number and reference it in your payment.

> Since all the scientists at Mythical University travel a lot, you have to repeat this tedious, error-prone and risky process several times per day. 

Reviewer: ```suggestion
Because the scientists at Mythical University travel frequently, you repeat this tedious, error-prone, and risky process many times a day. 
```

> You decide to get a bot to do this.   

Reviewer: ```suggestion
You decide to get a bot to do this work.
```

> The following example shows how you automate the retrieving of the order ID from a PDF and of the tracking number from the legacy system using MuleSoft RPA as an invocable automation step. How to download the right attachment before and how to proceed afterwards isn't part of this example.

Reviewer: ```suggestion
This example shows how to get the order ID and tracking number from the legacy system automatically. You can use a PDF and MuleSoft RPA as an invocable automation step to do this. The example doesn’t include downloading the attachment and any steps afterward.
```

## Design a Model of the Process in RPA Manager

In RPA Manager, you open a new project and design a model of the process.

The model contains the following activities:

. Extract the order ID from the order confirmation.
. Extract the corresponding tracking number from the legacy system.

> . Return either the tracking number or an information that it couldn't be found.

Reviewer: ```suggestion
. Return either the tracking number or information that it couldn’t be found.
```

> image::extractrackingnumber-bpmn.png["In the flow, activities one and two are sequential. Depending on whether the order id is found, an exclusive gateway branches the flow to the third activity."]

Reviewer: I think the important part of the image is the activities, not the description of the decision tree.

A flow showing extracting the order ID and then the tracking ID from the order confirmation...

Reviewer: Does this graphic follow Salesforce guidelines?

Author: The modelling of the decision tree is as important as the modelling of the activities. This is a screenshot of our product showing the model in BPMN standard (Business Process Modelling and Notation).

Author: I added that information together with a link to the paragraph.

> The user tasks in the third activity are for illustrational purposes and can be replaced by steps to prepare the returned data for processing in another automation step.

Reviewer: Can you make this active voice?

After you finish the first draft of the model, you move the project to the build phase.

Then, you open the project in RPA Builder to implement the workflows for the bot activities.

## Create Activity Parameters in RPA Builder

The process uses variables to store and route data. You define these _Activity Parameters_ in RPA Builder.

> For your process you need three Activity Parameters:

Reviewer: ```suggestion
For your process, you need three Activity Parameters:
```

* order_id

> ** an alphanumeric variable

Reviewer: ```suggestion
** is an alphanumeric variable
```

Reviewer: (to make the bullets parallel)

** stores the order ID found in the order confirmation
** is returned at the end of the process
* tracking_no

> ** an alphanumeric variable

Reviewer: ```suggestion
** is an alphanumeric variable
```

** stores the corresponding tracking number found in the order tracker application
** is returned at the end of the process if it is found
* is_order_id_found

> ** a boolean variable

Reviewer: ```suggestion
** is a boolean variable
```

** is set to true if the order id is found in the order tracking system
** routes the process flow through the gateway accordingly

image::activityparameters.png["The Activity Parameter tab in the Business Process Initialization window on the modelling workbench in RPA Builder shows the three parameters of the process."]

> To use Activity Parameters in a Workflow, move them from *Available Parameters* to *Used in this Workflow* in the *Activity Parameters* Action Step of the *Workflow Initialization* section. The parameters retain the values set in previous workflows throughout the process.

Reviewer: Should Workflow be lowercase?

Author: No, as Activity Parameter, Workflow is a proper noun in MuleSoft RPA.

Reviewer: Speaking of which, CrisV doesn't init cap Action Step. It would be good to get consistent on that. 

Our style is to favor lowercase (generic) unless it's a reference to something in the UI.

image::usecase-legacysystem-useactivityparameters.png[]

## Implement the Activities

You find some details of the implementation of the activities as _Workflows_ on the next pages:

> . xref:automation-usecase-legacysystemdataextraction-extractfrompdf.adoc[]

Reviewer: Are these sequential steps? 
If so, maybe the intro should be something like "The next steps in implementing activities as Workflows are...."

= Usecase: Extracting Data from a Legacy System
:page-pagination: next

> Extracting data from legacy systems is a common use case for RPA. Companies use legacy applications because they still work, they contain data crucial for daily work, and it's complicated and expensive to replace them. Companies also work with newer applications that use the same data. If the legacy application offers no way to transfer the data automatically, employees must copy and paste data sets manually. This takes up a lot of time, is error-prone, and tedious. Robots can extract the data faster, securer, and don't get bored. 

Reviewer 2: ```suggestion
Extracting data from legacy systems is a common use case for RPA. Companies use legacy applications because they still work, they contain data crucial for daily work, and it's complicated and expensive to replace them. Companies also work with newer applications that use the same data. If the legacy application offers no way to transfer the data automatically, employees must copy and paste data sets manually. This takes up a lot of time and is error-prone and tedious. Robots can extract the data faster and more securely and don't get bored. 
```

Examples of data extraction use cases are:

* Retrieve an order ID from a legagy system to get a tracking number from an order tracking application.
* Retrieve data from a legacy AMS to create a CRM report in Salesforce. (Check out this https://anypoint.mulesoft.com/exchange/org.mule.examples/mulesoft-accelerator-for-financial-services/minor/1.10/pages/604-0mq/Use%20case%207%20-%20Unlock%20AMS%20through%20hyperautomation%20for%20insurance%20brokers/[Accelerator]!)
* Migrate all of the data of a legacy application to a state-of-the-art application.

This example illustrates the following tasks:

* Copy the contents from a PDF file to a text file.
* Compare strings.
* Copy a substring from a text file.
* Find text on a screen using optical character recognition (OCR).
* Interact with a screen using mouse actions and keystrokes.
* Use activity parameters to transfer data and to determine which path to follow in the model.

== Example Scenario

> Mythical University orders a travel package for a field excursions. As an accountant, you receive an email and save the attached order confirmation to your computer.

Reviewer 2: ```suggestion
Mythical University orders a travel package for a field excursion. As an accountant, you receive an email and save the attached order confirmation to your computer.
```

An order confirmation from your travel agency Wonderlust Travel typically looks like this:

[[confirmationletter-image]]

> image::usecase-1-wonderlusttravels-orderconfirmation-pdf.png["Letter containing a logo, headers, text and tables. The order ID is contained in a table in a header in the middle of the page. The label says 'Order ID'.", 75%]

Reviewer 2: ```suggestion
image::usecase-1-wonderlusttravels-orderconfirmation-pdf.png["Letter containing a logo, headers, text, and tables. The order ID is contained in a table in a header in the middle of the page. The label says 'Order ID'.", 75%]
```

You need to refer the corresponding tracking number in your payment, so you search for the order ID in the order tracking system.

> Unfortunately, your order tracking system is old and you can access it only by typing in and reading from a terminal screen:

Reviewer 2: ```suggestion
Unfortunately, your order tracking system is old and can only be accessed by typing in and reading from a terminal screen:
```

[[trackingsystem-allentries-image]]

> image::as400ordertrackingscreen.png["Legacy order tracking system showing green glowing text in a terminal. Above a list of orders, there is an entry field for the order id labeled 'Order-ID' and a button labeled 'Search' next to it.", 75%]

Reviewer 2: ```suggestion
image::as400ordertrackingscreen.png["Legacy order tracking system showing green glowing text in a terminal. Above a list of orders, there is an entry field for the order ID labeled 'Order-ID' and a button labeled 'Search' next to it.", 75%]
```

Now you can copy the tracking number and reference it in your payment.

Because the scientists at Mythical University travel frequently, you repeat this tedious, error-prone, and risky process many times a day. 

You decide to get a bot to do this work.

This example shows how to get the order ID and tracking number from the legacy system automatically. You can use a PDF and MuleSoft RPA as an invocable automation step to do this. The example doesn’t include downloading the attachment and any steps afterward.

== Design a Model of the Process in RPA Manager

In RPA Manager, you open a new project and design a model of the process using the https://www.omg.org/spec/BPMN/2.0/[Business Process Model and Notation (BPMN) standard].

The model contains the following activities:

. Extract the order ID from the order confirmation.
. Extract the corresponding tracking number from the legacy system.
. Return either the tracking number or information that it couldn’t be found.

image::extractrackingnumber-bpmn.png["In the model flow, activities one and two are sequential. Depending on whether the order id is found, an exclusive gateway branches the flow to the third activity."]

> You include the user tasks in the third activity for illustrational and testing purposes and can later replace them by steps to prepare the returned data for processing in another automation step.

Reviewer 2: ```suggestion
You include the user tasks in the third activity for illustration and testing purposes and can later replace them with steps to prepare the returned data for processing in another automation step.
```
